File: java/Graph.java
import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.PriorityQueue;
import java.util.Set;

/**
 * Graph.java
 *
 * The transport network is stored as a GRAPH using an ADJACENCY LIST.
 *
 *   vertex  = a location  (Ahmedabad, Rajkot, Surat, ...)
 *   edge    = a road between two locations, weight = distance in km
 *
 * Internal storage:
 *   Map<String, ArrayList<Edge>> adjacencyList
 *      key   -> location name
 *      value -> list of edges starting from that location
 *
 * Why an adjacency list?
 *   - memory needed is O(V + E) instead of O(V*V) for a matrix
 *   - finding all neighbours of a city takes O(degree) only
 *   - it grows automatically when the user adds a new location
 *
 * This project uses a HashMap only as the container of the adjacency list
 * (it gives O(1) lookup of a vertex's edge list). No external library is used.
 */
public class Graph {

    private Map<String, ArrayList<Edge>> adjacencyList;
    private ArrayList<String> locations;      // keeps the insertion order

    public Graph() {
        adjacencyList = new HashMap<String, ArrayList<Edge>>();
        locations = new ArrayList<String>();
    }

    // =====================================================================
    //                              VERTICES
    // =====================================================================

    /**
     * Add a new location (vertex) to the graph.
     * Returns false when the name is invalid or the location already exists.
     */
    public boolean addLocation(String name) {
        if (name == null) {
            return false;
        }
        String key = name.trim();
        if (key.length() < 2) {                       // invalid input
            return false;
        }
        if (adjacencyList.containsKey(key)) {         // duplicate location
            return false;
        }
        adjacencyList.put(key, new ArrayList<Edge>());
        locations.add(key);
        return true;
    }

    /** Check whether a location exists. */
    public boolean contains(String city) {
        return adjacencyList.containsKey(city);
    }

    /** All locations, in the order they were added. */
    public ArrayList<String> getLocations() {
        return locations;
    }

    /** Number of vertices. */
    public int getLocationCount() {
        return locations.size();
    }

    /** Adjacency list (edge list) of one location. */
    public ArrayList<Edge> getNeighbours(String city) {
        ArrayList<Edge> edges = adjacencyList.get(city);
        if (edges == null) {
            return new ArrayList<Edge>();
        }
        return edges;
    }

    /** How many roads start from this location. */
    public int getDegree(String city) {
        return getNeighbours(city).size();
    }

    // =====================================================================
    //                                EDGES
    // =====================================================================

    /**
     * Add a route (weighted edge) between two existing locations.
     * Every road is two-way, so the edge is inserted in both directions.
     */
    public boolean addRoute(String from, String to, int distance) {
        if (!adjacencyList.containsKey(from) || !adjacencyList.containsKey(to)) {
            return false;                             // unknown location
        }
        if (from.equals(to) || distance <= 0) {
            return false;                             // invalid route
        }
        adjacencyList.get(from).add(new Edge(to, distance));
        adjacencyList.get(to).add(new Edge(from, distance));
        return true;
    }

    /** Remove the road between two locations (both directions). */
    public boolean removeRoute(String from, String to) {
        if (!adjacencyList.containsKey(from) || !adjacencyList.containsKey(to)) {
            return false;
        }
        boolean removed = removeOneDirection(from, to);
        removeOneDirection(to, from);
        return removed;
    }

    private boolean removeOneDirection(String from, String to) {
        ArrayList<Edge> edges = adjacencyList.get(from);
        for (int i = 0; i < edges.size(); i++) {
            if (edges.get(i).getCity().equals(to)) {
                edges.remove(i);
                return true;
            }
        }
        return false;
    }

    /**
     * Every road exactly once.
     * Because a two-way road is stored twice, we only keep the pair where
     * the first city name is alphabetically smaller.
     */
    public ArrayList<Route> getUniqueRoutes() {
        ArrayList<Route> routes = new ArrayList<Route>();
        for (String city : locations) {
            for (Edge edge : getNeighbours(city)) {
                if (city.compareTo(edge.getCity()) < 0) {
                    routes.add(new Route(city, edge.getCity(), edge.getDistance()));
                }
            }
        }
        return routes;
    }

    /** Number of roads (edges) counted only once. */
    public int getRouteCount() {
        return getUniqueRoutes().size();
    }

    /** Total length of all roads, in kilometres. */
    public int getTotalNetworkDistance() {
        int total = 0;
        for (Route route : getUniqueRoutes()) {
            total = total + route.getDistance();
        }
        return total;
    }

    // =====================================================================
    //                         DISPLAY (menu option 3)
    // =====================================================================

    public void display() {
        System.out.println();
        System.out.println("============ TRANSPORT NETWORK (ADJACENCY LIST) ============");
        if (locations.isEmpty()) {
            System.out.println("   (no location added yet)");
        }
        for (String city : locations) {
            System.out.print("   " + city + " -> ");
            ArrayList<Edge> edges = getNeighbours(city);
            if (edges.isEmpty()) {
                System.out.println("null");
            } else {
                for (int i = 0; i < edges.size(); i++) {
                    System.out.print(edges.get(i).getCity()
                            + " (" + edges.get(i).getDistance() + " km)");
                    if (i < edges.size() - 1) {
                        System.out.print(", ");
                    }
                }
                System.out.println();
            }
        }
        System.out.println("============================================================");
        System.out.println("   Locations : " + getLocationCount());
        System.out.println("   Routes    : " + getRouteCount());
        System.out.println("   Network   : " + getTotalNetworkDistance() + " km of road");
        System.out.println("============================================================");
    }

    // =====================================================================
    //               DIJKSTRA'S SHORTEST PATH (menu options 4 and 5)
    // =====================================================================

    /**
     * DIJKSTRA'S ALGORITHM (greedy) using a min-heap priority queue.
     *
     * 1. dist[city] = infinity for every city, dist[source] = 0
     * 2. push the source into the priority queue with key 0
     * 3. take out the closest city that is not finalised yet (greedy choice)
     * 4. for every neighbour compute  candidate = dist[current] + weight
     * 5. if candidate < dist[neighbour]  ->  RELAX the edge (update dist and
     *    previous) and push the neighbour with its new smaller distance
     * 6. repeat until the destination is finalised or the queue is empty
     * 7. rebuild the route by walking backwards through previous[]
     *
     * Time complexity : O((V + E) log V)
     * Space complexity: O(V)
     */
    public DijkstraResult shortestPath(String source, String destination) {

        ArrayList<String> emptyPath = new ArrayList<String>();
        ArrayList<String> emptyOrder = new ArrayList<String>();

        if (!adjacencyList.containsKey(source) || !adjacencyList.containsKey(destination)) {
            return new DijkstraResult(emptyPath, -1, emptyOrder);
        }

        Map<String, Integer> distance = new HashMap<String, Integer>();
        Map<String, String> previous = new HashMap<String, String>();
        Set<String> visited = new HashSet<String>();
        ArrayList<String> visitedOrder = new ArrayList<String>();

        // step 1 : initialise every distance to "infinity"
        for (String city : locations) {
            distance.put(city, Integer.MAX_VALUE);
            previous.put(city, null);
        }
        distance.put(source, 0);

        // The java.util.PriorityQueue is a BINARY MIN-HEAP, so the Edge with
        // the smallest distance is always at the root -> O(log n) operations.
        PriorityQueue<Edge> priorityQueue = new PriorityQueue<Edge>(new Comparator<Edge>() {
            public int compare(Edge first, Edge second) {
                return first.getDistance() - second.getDistance();
            }
        });
        priorityQueue.add(new Edge(source, 0));

        while (!priorityQueue.isEmpty()) {
            Edge current = priorityQueue.poll();          // step 3

            if (visited.contains(current.getCity())) {
                continue;                                 // already finalised
            }
            visited.add(current.getCity());
            visitedOrder.add(current.getCity());

            if (current.getCity().equals(destination)) {
                break;                                    // answer is final
            }

            for (Edge edge : getNeighbours(current.getCity())) {
                if (visited.contains(edge.getCity())) {
                    continue;
                }
                int candidate = distance.get(current.getCity()) + edge.getDistance();

                if (candidate < distance.get(edge.getCity())) {     // relaxation
                    distance.put(edge.getCity(), candidate);
                    previous.put(edge.getCity(), current.getCity());
                    priorityQueue.add(new Edge(edge.getCity(), candidate));
                }
            }
        }

        // no route at all -> report it instead of printing a wrong path
        if (distance.get(destination) == Integer.MAX_VALUE) {
            return new DijkstraResult(emptyPath, -1, visitedOrder);
        }

        // step 7 : rebuild the path from destination back to source
        ArrayList<String> path = new ArrayList<String>();
        String step = destination;
        while (step != null) {
            path.add(0, step);
            step = previous.get(step);
        }
        return new DijkstraResult(path, distance.get(destination), visitedOrder);
    }

    // =====================================================================
    //     BFS (kept only to compare it with Dijkstra during the viva)
    // =====================================================================

    /**
     * Breadth First Search. It ignores the weights and returns the route with
     * the SMALLEST NUMBER OF CITIES, which is not always the shortest route.
     */
    public ArrayList<String> breadthFirstSearch(String source, String destination) {

        ArrayList<String> path = new ArrayList<String>();
        if (!adjacencyList.containsKey(source) || !adjacencyList.containsKey(destination)) {
            return path;
        }

        Set<String> visited = new HashSet<String>();
        Map<String, String> previous = new HashMap<String, String>();
        ArrayList<String> queue = new ArrayList<String>();     // simple FIFO queue

        visited.add(source);
        previous.put(source, null);
        queue.add(source);

        while (!queue.isEmpty()) {
            String current = queue.remove(0);
            if (current.equals(destination)) {
                break;
            }
            for (Edge edge : getNeighbours(current)) {
                if (!visited.contains(edge.getCity())) {
                    visited.add(edge.getCity());
                    previous.put(edge.getCity(), current);
                    queue.add(edge.getCity());
                }
            }
        }

        if (!visited.contains(destination)) {
            return path;
        }
        String step = destination;
        while (step != null) {
            path.add(0, step);
            step = previous.get(step);
        }
        return path;
    }

    /** Total distance of an already known path (-1 if a road is missing). */
    public int getPathDistance(ArrayList<String> path) {
        int total = 0;
        for (int i = 0; i < path.size() - 1; i++) {
            boolean found = false;
            for (Edge edge : getNeighbours(path.get(i))) {
                if (edge.getCity().equals(path.get(i + 1))) {
                    total = total + edge.getDistance();
                    found = true;
                    break;
                }
            }
            if (!found) {
                return -1;
            }
        }
        return total;
    }
}

File: src/lib/ds/Graph.ts
import { PriorityQueue } from "./PriorityQueue";

/** A directed weighted edge of the graph (an entry of an adjacency list) */
export interface GraphEdge {
  to: string;
  weight: number;
}

export interface RouteStep {
  visited: string;
  relaxed: string;
  newDistance: number;
  improved: boolean;
}

export interface DijkstraResult {
  reachable: boolean;
  path: string[];
  distance: number;
  visitedOrder: string[];
  distances: Record<string, number>;
  steps: RouteStep[];
}

/**
 * GRAPH using an ADJACENCY LIST.
 * Key = location name, Value = array of { to, weight } edges.
 * Space: O(V + E). Finding all neighbours of a vertex: O(degree).
 */
export class Graph {
  private adjacency = new Map<string, GraphEdge[]>();

  /** Add a location (vertex). Returns false if it already exists. */
  addLocation(name: string): boolean {
    const key = name.trim();
    if (key.length === 0 || this.adjacency.has(key)) return false;
    this.adjacency.set(key, []);
    return true;
  }

  hasLocation(name: string): boolean {
    return this.adjacency.has(name);
  }

  /** Add a route (edge). Two-way roads are stored as two directed edges. */
  addRoute(from: string, to: string, weight: number, bidirectional = true): boolean {
    const a = from.trim();
    const b = to.trim();
    if (!this.adjacency.has(a) || !this.adjacency.has(b)) return false;
    if (weight <= 0) return false;
    this.adjacency.get(a)!.push({ to: b, weight });
    if (bidirectional) this.adjacency.get(b)!.push({ to: a, weight });
    return true;
  }

  removeRoute(from: string, to: string): boolean {
    const a = from.trim();
    const b = to.trim();
    if (!this.adjacency.has(a) || !this.adjacency.has(b)) return false;
    const outA = this.adjacency.get(a)!;
    const idxA = outA.findIndex((e) => e.to === b);
    if (idxA >= 0) outA.splice(idxA, 1);
    const outB = this.adjacency.get(b)!;
    const idxB = outB.findIndex((e) => e.to === a);
    if (idxB >= 0) outB.splice(idxB, 1);
    return idxA >= 0 || idxB >= 0;
  }

  locations(): string[] {
    return [...this.adjacency.keys()];
  }

  neighbors(name: string): GraphEdge[] {
    return this.adjacency.get(name) ?? [];
  }

  degree(name: string): number {
    return this.neighbors(name).length;
  }

  routes(): { from: string; to: string; weight: number }[] {
    const out: { from: string; to: string; weight: number }[] = [];
    for (const [from, edges] of this.adjacency) {
      for (const e of edges) out.push({ from, to: e.to, weight: e.weight });
    }
    return out;
  }

  uniqueRoutes(): { from: string; to: string; weight: number }[] {
    const seen = new Set<string>();
    const out: { from: string; to: string; weight: number }[] = [];
    for (const [from, edges] of this.adjacency) {
      for (const e of edges) {
        const key = [from, e.to].sort().join("~");
        if (seen.has(key)) continue;
        seen.add(key);
        out.push({ from, to: e.to, weight: e.weight });
      }
    }
    return out;
  }

  size(): number {
    return this.adjacency.size;
  }

  /**
   * DIJKSTRA'S SHORTEST PATH ALGORITHM (greedy).
   * Uses a min-heap priority queue -> O((V + E) log V).
   * Every step is recorded so the UI can explain the relaxation process.
   */
  dijkstra(source: string, destination: string): DijkstraResult {
    const distances: Record<string, number> = {};
    const previous: Record<string, string | null> = {};
    const visited = new Set<string>();
    const visitedOrder: string[] = [];
    const steps: RouteStep[] = [];
    const pq = new PriorityQueue<{ name: string; dist: number }>((a, b) => a.dist - b.dist);

    for (const city of this.adjacency.keys()) {
      distances[city] = Infinity;
      previous[city] = null;
    }
    if (!this.adjacency.has(source) || !this.adjacency.has(destination)) {
      return { reachable: false, path: [], distance: Infinity, visitedOrder: [], distances, steps };
    }
    distances[source] = 0;
    pq.enqueue({ name: source, dist: 0 });

    while (!pq.isEmpty()) {
      const current = pq.dequeue()!;
      if (visited.has(current.name)) continue;
      visited.add(current.name);
      visitedOrder.push(current.name);

      if (current.name === destination) break;

      for (const edge of this.neighbors(current.name)) {
        if (visited.has(edge.to)) continue;
        const candidate = distances[current.name] + edge.weight;
        const improved = candidate < distances[edge.to];
        steps.push({
          visited: current.name,
          relaxed: edge.to,
          newDistance: candidate,
          improved,
        });
        if (improved) {
          distances[edge.to] = candidate;
          previous[edge.to] = current.name;
          pq.enqueue({ name: edge.to, dist: candidate });
        }
      }
    }

    if (distances[destination] === Infinity) {
      return { reachable: false, path: [], distance: Infinity, visitedOrder, distances, steps };
    }

    const path: string[] = [];
    let walker: string | null = destination;
    while (walker !== null) {
      path.unshift(walker);
      walker = previous[walker];
    }
    return { reachable: true, path, distance: distances[destination], visitedOrder, distances, steps };
  }

  /** Breadth First Search — fewest number of stops (ignores distance). Used for the viva comparison. */
  bfs(source: string, destination: string): string[] {
    if (!this.adjacency.has(source) || !this.adjacency.has(destination)) return [];
    const visited = new Set<string>([source]);
    const previous: Record<string, string | null> = { [source]: null };
    const queue: string[] = [source];
    while (queue.length > 0) {
      const current = queue.shift() as string;
      if (current === destination) break;
      for (const edge of this.neighbors(current)) {
        if (!visited.has(edge.to)) {
          visited.add(edge.to);
          previous[edge.to] = current;
          queue.push(edge.to);
        }
      }
    }
    if (!visited.has(destination)) return [];
    const path: string[] = [];
    let walker: string | null = destination;
    while (walker !== null) {
      path.unshift(walker);
      walker = previous[walker];
    }
    return path;
  }

  pathDistance(path: string[]): number {
    let total = 0;
    for (let i = 0; i < path.length - 1; i++) {
      const edge = this.neighbors(path[i]).find((e) => e.to === path[i + 1]);
      if (!edge) return -1;
      total += edge.weight;
    }
    return total;
  }
}
